storage/database.go

```diff
@@ -54,45 +55,74 @@ func (s *dbRepoStore) GetByEndpoints(endpoints ...string) ([]*model.Repository,
 
 func (s *dbRepoStore) SetStatus(repo *model.Repository, status model.FetchStatus) error {
 	repo.Status = status
-	_, err := s.RepositoryStore.Update(
+	return s.updateWithRefsChanged(
 		repo,
 		model.Schema.Repository.Status,
 	)
-	return err
 }
 
 func (s *dbRepoStore) SetEndpoints(repo *model.Repository, endpoints ...string) error {
 	repo.Endpoints = endpoints
-	_, err := s.Update(repo, model.Schema.Repository.Endpoints)
-	return err
+	return s.updateWithRefsChanged(repo, model.Schema.Repository.Endpoints)
 }
 
 func (s *dbRepoStore) UpdateFailed(repo *model.Repository, status model.FetchStatus) error {
 	repo.Status = status
-	_, err := s.Update(repo,
+	return s.updateWithRefsChanged(repo,
 		model.Schema.Repository.UpdatedAt,
 		model.Schema.Repository.FetchErrorAt,
-		model.Schema.Repository.References,
 		model.Schema.Repository.Status,
 	)
-
-	return err
 }
 
 func (s *dbRepoStore) UpdateFetched(repo *model.Repository, fetchedAt time.Time) error {
 	repo.Status = model.Fetched
 	repo.FetchedAt = &fetchedAt
 	repo.LastCommitAt = lastCommitTime(repo.References)
 
-	_, err := s.Update(repo,
+	return s.updateWithRefsChanged(repo,
 		model.Schema.Repository.UpdatedAt,
 		model.Schema.Repository.FetchedAt,
 		model.Schema.Repository.LastCommitAt,
 		model.Schema.Repository.Status,
-		model.Schema.Repository.References,
 	)
+}
 
-	return err
+func (s *dbRepoStore) updateWithRefsChanged(repo *model.Repository, fields ...kallax.SchemaField) error {
+	return s.Transaction(func(store *model.RepositoryStore) error {
+		var refStore model.ReferenceStore
+		kallax.StoreFrom(&refStore, store)
+
+		refs, err := refStore.FindAll(model.NewReferenceQuery().FindByRepository(repo.ID))
+		if err != nil {
+			return err
+		}
+
+		for _, ref := range refs {
+			if err := refStore.Delete(ref); err != nil {
```

Comment:
instead of diffing for changes, which can get hairy, I decided to just delete+create

Reply:
But this is not atomic, right? Should be in a transaction.

Reply:
It's in a transaction https://github.com/src-d/borges/pull/210/files/0aa0817bbdde008381ca0d866f4739664b06c2b2#diff-97b2ac718f2c9a1379f7f9d43fb9dc34R92